internal/website/gatherexamples/gatherexamples.go

```diff
@@ -0,0 +1,271 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Command gatherexamples extracts all examples in a Go module into a JSON
+// object. This is used as input for building the Go CDK Hugo website.
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/printer"
+	"go/types"
+	"os"
+	"sort"
+	"strings"
+
+	"golang.org/x/tools/go/packages"
+)
+
+func main() {
+	flag.Usage = func() {
+		out := flag.CommandLine.Output()
+		fmt.Fprintln(out, "usage: gatherexamples [options] DIR [...]")
+		fmt.Fprintln(out)
+		fmt.Fprintln(out, "Options:")
+		flag.PrintDefaults()
+	}
+	outputPath := flag.String("o", "", "path to output file (default to stdout)")
+	pattern := flag.String("pattern", "./...", "pattern to use at each directory")
+	flag.Parse()
+	if flag.NArg() == 0 {
+		flag.Usage()
+		os.Exit(2) // matches with flag package
+	}
+
+	// Load packages in each module named on the command line and find
+	// all examples.
+	allExamples := make(map[string]string)
+	for _, dir := range flag.Args() {
+		cfg := &packages.Config{
+			Mode:  gatherLoadMode,
+			Dir:   dir,
+			Tests: true,
+		}
+		pkgs, err := packages.Load(cfg, *pattern)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: load %s: %v\n", dir, err)
+			os.Exit(1)
+		}
+		for exampleName, source := range gather(pkgs) {
+			allExamples[exampleName] = source
+		}
+	}
+
+	// Write all examples as a JSON object.
+	data, err := json.MarshalIndent(allExamples, "", "\t")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: generate JSON: %v\n", err)
+		os.Exit(1)
+	}
+	data = append(data, '\n')
+
+	output := os.Stdout
+	if *outputPath != "" && *outputPath != "-" {
+		var err error
+		output, err = os.Create(*outputPath)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: %v\n", err)
+			os.Exit(1)
+		}
+		defer func() {
+			if err := output.Close(); err != nil {
+				fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+				os.Exit(1)
+			}
+		}()
+	}
+	if _, err := output.Write(data); err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+const gatherLoadMode packages.LoadMode = packages.NeedName |
+	packages.NeedFiles |
+	packages.NeedTypes |
+	packages.NeedSyntax |
+	packages.NeedTypesInfo
+
+func gather(pkgs []*packages.Package) map[string]string {
+	examples := make(map[string]string)
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			for _, decl := range file.Decls {
+				// Determine whether this declaration is an example function.
+				fn, ok := decl.(*ast.FuncDecl)
+				if !ok {
+					continue
+				}
+				if !ok || !strings.HasPrefix(fn.Name.Name, "Example") || len(fn.Type.Params.List) > 0 || len(fn.Type.Params.List) > 0 {
+					continue
+				}
+
```

Comment:
As suggested in team meeting, consider adding a tag/comment of some kind that an example must
have in order to be included.

This
a) makes the data file only include the examples the website needs, reducing churn.
b) makes it obvious in the example source that it's being used in this way.

Reply:
Done.